File: SenLib/Sen3/FileFixes/insa09.cs
using HyoutaUtils;
using HyoutaUtils.Streams;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SenLib.Sen3.FileFixes {
	class insa09 : FileMod {
		public string GetDescription() {
			return "Fix duplicate Chapter 3 map travel sequence.";
		}

		public IEnumerable<FileModResult> TryApply(FileStorage storage) {
			var file = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x97b37cb0d324014dul, 0x8db6b2965f1836eful, 0xfdb1ce01u));
			if (file == null) {
				return null;
			}

			var ms = new System.IO.MemoryStream();
			HyoutaUtils.Bps.BpsPatcher.ApplyPatchToStream(file, new DuplicatableByteArrayStream(Properties.Resources.insa_09), ms);
			ms.Position = 0;
			return new FileModResult[] { new FileModResult("data/movie_us/webm/insa_09.webm", ms) };
		}

		public IEnumerable<FileModResult> TryRevert(FileStorage storage) {
			var file = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x97b37cb0d324014dul, 0x8db6b2965f1836eful, 0xfdb1ce01u));
			if (file == null) {
				return null;
			}
			return new FileModResult[] { new FileModResult("data/movie_us/webm/insa_09.webm", file) };
		}
	}
}

File: native/sen3/FileFixes/insa09.cpp
#include <filesystem>
#include <string_view>
#include <vector>

#include "../file_fixes.h"

#include "../../bps.h"
#include "../../p3a/pack.h"
#include "../../p3a/structs.h"
#include "../../sha1.h"
#include "../../util/stream.h"

namespace {
static constexpr char PatchData[] = {
#include "embed_sen3_insa_09.h"
};
static constexpr size_t PatchLength = sizeof(PatchData);
} // namespace

namespace SenLib::Sen3::FileFixes::insa09 {
std::string_view GetDescription() {
    return "Fix duplicate Chapter 3 map travel sequence.";
}

bool TryApply(const SenPatcher::GetCheckedFileCallback& getCheckedFile,
              std::vector<SenPatcher::P3APackFile>& result) {
    try {
        auto file = getCheckedFile(
            "data/movie_us/webm/insa_09.webm",
            9103875,
            SenPatcher::SHA1FromHexString("97b37cb0d324014d8db6b2965f1836effdb1ce01"));
        if (!file) {
            return false;
        }

        auto& bin = file->Data;

        SenLib::DuplicatableByteArrayStream source(bin.data(), bin.size());
        SenLib::DuplicatableByteArrayStream patch(PatchData, PatchLength);
        std::vector<char> target;
        HyoutaUtils::Bps::ApplyPatchToStream(source, patch, target);

        // FIXME: update output filename here

        result.emplace_back(SenPatcher::P3APackFile{
            std::move(target), file->Filename, SenPatcher::P3ACompressionType::None});

        return true;
    } catch (...) {
        return false;
    }
}
} // namespace SenLib::Sen3::FileFixes::insa09
